ra/ra.go

```diff
@@ -1563,24 +1569,54 @@ func (ra *RegistrationAuthorityImpl) NewOrder(ctx context.Context, req *rapb.New
 		return nil, err
 	}
 
-	gotAuthzFor := make(map[string]bool, len(order.Names))
+	// Collect up the authorizations we found into a map keyed by the domains the
+	// authorizations correspond to
+	nameToExistingAuthz := make(map[string]*corepb.Authorization, len(orderIdentifiers))
 	for _, v := range existingAuthz.Authz {
-		gotAuthzFor[*v.Domain] = true
-		order.Authorizations = append(order.Authorizations, *v.Authz.Id)
+		nameToExistingAuthz[*v.Domain] = v.Authz
+	}
+
+	// For each of the names in the order, if there is an acceptable
+	// existing authz, append its ID to the order to reuse it. Otherwise track
+	// that there is a missing authz for a name
+	var missingAuthzs []core.AcmeIdentifier
+	for i, name := range order.Names {
+		ident := orderIdentifiers[i]
+		// If there is an existing authz, process it
+		if authz, exists := nameToExistingAuthz[name]; exists {
```

Comment:
style nit: since this is the larger block I'd switch the `exists` to `!exists` and put the `missingAuthzs = append(missingAuthzs, ident); continue` in this `if`, but ¯\_(ツ)_/¯.

Reply:
Resolved in 8606058